test/parallel/test-runner-run-watch.mjs

```diff
@@ -0,0 +1,187 @@
+}
+
+const runner = join(import.meta.dirname, '..', 'fixtures', 'test-runner-watch.mjs');
+
+async function testWatch({ fileToUpdate, file, action = 'update', cwd = tmpdir.path }) {
+  const ran1 = util.createDeferredPromise();
+  const ran2 = util.createDeferredPromise();
+  const args = [runner];
+  if (file) args.push('--file', file);
+  const child = spawn(process.execPath,
+                      args,
+                      { encoding: 'utf8', stdio: 'pipe', cwd });
+  let stdout = '';
+  let currentRun = '';
+  const runs = [];
+
+  child.stdout.on('data', (data) => {
+    stdout += data.toString();
+    currentRun += data.toString();
+    const testRuns = stdout.match(/# duration_ms\s\d+/g);
+    if (testRuns?.length >= 1) ran1.resolve();
+    if (testRuns?.length >= 2) ran2.resolve();
+  });
+
+  const testUpdate = async () => {
+    await ran1.promise;
+    const content = fixtureContent[fileToUpdate];
+    const path = fixturePaths[fileToUpdate];
+    const interval = setInterval(() => writeFileSync(path, content), common.platformTimeout(1000));
+    await ran2.promise;
+    runs.push(currentRun);
+    clearInterval(interval);
+    child.kill();
+    for (const run of runs) {
+      assert.doesNotMatch(run, /run\(\) is being called recursively/);
+      assert.match(run, /# tests 1/);
+      assert.match(run, /# pass 1/);
+      assert.match(run, /# fail 0/);
+      assert.match(run, /# cancelled 0/);
+    }
+  };
+
+  const testRename = async () => {
+    await ran1.promise;
+    const fileToRenamePath = tmpdir.resolve(fileToUpdate);
+    const newFileNamePath = tmpdir.resolve(`test-renamed-${fileToUpdate}`);
+    const interval = setInterval(() => renameSync(fileToRenamePath, newFileNamePath), common.platformTimeout(1000));
+    await ran2.promise;
+    runs.push(currentRun);
+    clearInterval(interval);
+    child.kill();
+
+    for (const run of runs) {
+      assert.doesNotMatch(run, /run\(\) is being called recursively/);
+      assert.doesNotMatch(run, /MODULE_NOT_FOUND/);
+      assert.match(run, /# tests 1/);
+      assert.match(run, /# pass 1/);
+      assert.match(run, /# fail 0/);
+      assert.match(run, /# cancelled 0/);
+    }
+  };
+
+  const testRename2 = async () => {
+    await ran1.promise;
+    const fileToRenamePath = tmpdir.resolve(fileToUpdate);
+    const newFileNamePath = tmpdir.resolve(`test-renamed-${fileToUpdate}`);
+    const interval = setInterval(() => renameSync(fileToRenamePath, newFileNamePath), common.platformTimeout(1000));
+    await ran2.promise;
+    runs.push(currentRun);
+    clearInterval(interval);
+    child.kill();
+
+    for (const run of runs) {
+      assert.doesNotMatch(run, /run\(\) is being called recursively/);
+      assert.match(run, /# tests 1/);
+      assert.match(run, /# pass 1/);
+      assert.match(run, /# fail 0/);
+      assert.match(run, /# cancelled 0/);
+      assert.match(run, /MODULE_NOT_FOUND/);
+    }
+  };
```

Comment:
Can this be 1 function with an argument?
This is repeating the entire function for 1 `match`/`doesNotMatch`
```suggestion
  const testRename = async () => {
    await ran1.promise;
    const fileToRenamePath = tmpdir.resolve(fileToUpdate);
    const newFileNamePath = tmpdir.resolve(`test-renamed-${fileToUpdate}`);
    const interval = setInterval(() => renameSync(fileToRenamePath, newFileNamePath), common.platformTimeout(1000));
    await ran2.promise;
    runs.push(currentRun);
    clearInterval(interval);
    child.kill();

    for (const run of runs) {
      assert.doesNotMatch(run, /run\(\) is being called recursively/);
      if (action === 'rename2') {
        assert.match(run, /MODULE_NOT_FOUND/);
      } else {
        assert.doesNotMatch(run, /MODULE_NOT_FOUND/);
      }
      assert.match(run, /# tests 1/);
      assert.match(run, /# pass 1/);
      assert.match(run, /# fail 0/);
      assert.match(run, /# cancelled 0/);
    }
  };
```